### PR TITLE
Add support of clientSidePagination to addrbar extension

### PR DESCRIPTION
**Pull request by kerog**
**Bug fix?**
no

**New Feature?**
yes

**Resolve an issue?**
no

**Example(s)?**
I can't add example because editor haven't address bar.

### CODE REVIEW

File: src/extensions/addrbar/bootstrap-table-addrbar.js [hunk 1]
```diff
@@ -63,6 +63,23 @@ function _buildUrl (dict, url = window.location.search) {
...
 
+/*
+* function: _updateHistoryState
+* var _prefix = this.options.addrPrefix || ''
+* _updateHistoryState(_prefix)
+* returns void
+*/
+function _updateHistoryState ( _prefix ) {
+  const params = {}
+  params[`${_prefix}page`] = this.options.pageNumber,
```
Comment (UtechtDustin) [inline src/extensions/addrbar/bootstrap-table-addrbar.js:74]: That cant work, because of `this`.

Comment (UtechtDustin) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:74]: Add `this` as parameter then it should work.
Its not the best way to solve it, but the extension must be rewrite anyways.

Comment (UtechtDustin) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:74]: It also would be greate to check if `order`, `sort` and/or `search` are `empty` or `undefined`.
If that is the case we should not add them as url parameter (`order=undefined` looks bad).

**Comment by UtechtDustin** [comment]
https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/addrbar/bootstrap-table-addrbar.js#L119-L121

This part will fail with the page size (because the pageSize is selectable), which means the url parameter is useless.

**Comment by kerog** [comment]
> 
> 
> https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/addrbar/bootstrap-table-addrbar.js#L119-L121
> 
> This part will fail with the page size (because the pageSize is selectable), which means the url parameter is useless.

Only reason for keep this, is share full url between 2 users and initial set the bootstrap-table, of parameters like search, page, limit, etc.

**Comment by UtechtDustin** [comment]
> Only reason for keep this, is share full url between 2 users and initial set the bootstrap-table, of parameters like search, page, limit, etc.

I think you missunderstand what i mean.
I dont mean to remove a url parameter, but we need a other solution for the pageSize/size parameter because currently you cant use them.

**Comment by kerog** [comment]
> 
> 
> > Only reason for keep this, is share full url between 2 users and initial set the bootstrap-table, of parameters like search, page, limit, etc.
> 
> I think you missunderstand what i mean.
> I dont mean to remove a url parameter, but we need a other solution for the pageSize/size parameter because currently you cant use them.

Ok.

File: src/extensions/addrbar/bootstrap-table-addrbar.js (hunk 1 above)
Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js]: I've change call of this function and removes 'undefined' from urls

File: src/extensions/addrbar/bootstrap-table-addrbar.js [hunk 2]
```diff
@@ -86,25 +105,26 @@ $.BootstrapTable = class extends $.BootstrapTable {
...
 
         if (_onLoadSuccess) {
           _onLoadSuccess.call(this, data)
         }
       }
+      this.options.onPageChange = (number,size) => {
+        _updateHistoryState(this,_prefix)
+
+        if (_onPageChange) {
+          _onPageChange.call(this,number,size)
```
Comment (djhvscf) [inline src/extensions/addrbar/bootstrap-table-addrbar.js:125]: I think we can use apply in order to send all arguments.

Something like yourFunction.apply(...args)

File: src/extensions/addrbar/bootstrap-table-addrbar.js [hunk 3]
```diff
@@ -63,6 +65,24 @@ function _buildUrl (dict, url = window.location.search) {
...
+* returns void
+*/
+function _updateHistoryState ( table,_prefix ) {
+  const params = {}
+  params[`${_prefix}page`] = table.options.pageNumber,
+  params[`${_prefix}size`] = table.options.pageSize,
+  params[`${_prefix}order`] = table.options.sortOrder,
+  params[`${_prefix}sort`] = table.options.sortName,
+  params[`${_prefix}search`] = table.options.searchText
+  // h5提供的修改浏览器地址栏的方法
```
[the remark this answers is not in the document]
Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:82]: I only extract this part of code to new function, to remove duplicates function call.

[the remark this answers is not in the document]
Comment (UtechtDustin) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:82]: You can delete it, this comment is not needed (used a translator).

File: src/extensions/addrbar/bootstrap-table-addrbar.js (hunk 2 above)
Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:125]: I do it, like in L104-106.

File: src/extensions/addrbar/bootstrap-table-addrbar.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js]: Ok I remove it.

File: src/extensions/addrbar/bootstrap-table-addrbar.js [hunk 4]
```diff
@@ -44,6 +44,8 @@ function _buildUrl (dict, url = window.location.search) {
     // 搜索name=general这种形式的字符串(&是分隔符)
     const pattern = `${key}=([^&]*)`
     const targetStr = `${key}=${val}`
+    if (targetStr.includes('undefined'))
```
Comment (UtechtDustin) [inline src/extensions/addrbar/bootstrap-table-addrbar.js:47]: `typeof targetStr === undefined`

Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:47]: This don't works because targetStr is string like this
```
sort=undefined
```

Comment (UtechtDustin) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:47]: oh your right, then `val === undefined` should work i guess.

Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js]: Ok, I change this line.

File: src/extensions/addrbar/bootstrap-table-addrbar.js (hunk 2 above)
Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js]: Ok, I changed this.

File: src/extensions/addrbar/bootstrap-table-addrbar.js [hunk 5]
```diff
@@ -63,6 +65,23 @@ function _buildUrl (dict, url = window.location.search) {
...
 }
 
+/*
+* function: _updateHistoryState
+* var _prefix = this.options.addrPrefix || ''
+* var table = this
+* _updateHistoryState( table,_prefix)
+* returns void
+*/
+function _updateHistoryState ( table,_prefix ) {
```
Comment (wenzhixin) [inline src/extensions/addrbar/bootstrap-table-addrbar.js:75]: I think we can move this function into BootstrapTable class.

Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:75]: This function use _buildUrl from L42-L66

Comment (UtechtDustin) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:75]: @wenzhixin do you mean the whole extension or just this function ?
What would be improvement if we would add this to the core (if you dont mean the whole extension)

Comment (kerog) [reply src/extensions/addrbar/bootstrap-table-addrbar.js:75]: I mean only the function _buildUrl is using by _updateHistoryState, and if you plan move function _updateHistoryState you must also move _buildUrl.

File: src/extensions/addrbar/bootstrap-table-addrbar.js [hunk 6]
```diff
@@ -44,6 +44,8 @@ function _buildUrl (dict, url = window.location.search) {
     // 搜索name=general这种形式的字符串(&是分隔符)
     const pattern = `${key}=([^&]*)`
     const targetStr = `${key}=${val}`
+    if (val === undefined )
```
Comment (UtechtDustin) [inline src/extensions/addrbar/bootstrap-table-addrbar.js:47]: Could be changed to the normal syntax with `{}`, but as the extension should be rewritten someday its ok.

**Comment by djhvscf** [comment]
@wenzhixin @kerog any update here?

**Comment by kerog** [comment]
> 
> 
> @wenzhixin @kerog any update here?

Hi I'm waiting for your accept of changes (I change 'call' to 'action')

**Comment by djhvscf** [comment]
Ok, please resolve the conversations first

**Comment by kerog** [comment]
done

**Comment by wenzhixin** [comment]
Merged into develop: https://github.com/wenzhixin/bootstrap-table/commits/develop

**Pull request closed without merging**